scio-extra/src/main/scala/com/spotify/scio/extra/Sparkey.scala

```diff
@@ -0,0 +1,181 @@
+   * a base path representing two files - <sparkey-prefix>.spi and <sparkey-prefix>.spl.
+   */
+  trait SparkeyUri {
+    val basePath: String
+    def getReader(): SparkeyReader
+    override def toString(): String = basePath
+  }
+
+  object SparkeyUri {
+    def apply(path: String): SparkeyUri =
+      if (ScioUtil.isGcsUri(new URI(path))) new GcsSparkeyUri(path) else new LocalSparkeyUri(path)
+  }
+
+  implicit class SparkeyScioContext(val self: ScioContext) {
+    def sparkeySideInput(url: SparkeyUri): SideInput[SparkeyReader] = {
+      val view = self.parallelize(Seq(url)).applyInternal(View.asSingleton())
+      new SparkeySideInput(view)
+    }
+  }
+
+  implicit class SCollectionWithSparkeyWriter[K, V](val self: SCollection[(K, V)])
+                                                   (implicit ev1: K <:< String, ev2: V <:< String) {
+    /**
+     * Write the contents of this SCollection as a Sparkey file, either locally or on GCS.
+     *
+     * @return A singleton SCollection containing the [[SparkeyUri]] of the saved files.
+     */
+    def asSparkey(uri: SparkeyUri): SCollection[SparkeyUri] = self.transform { in =>
+      in.groupBy(_ => ())
+        .map { case (_, iter) =>
+          val writer = new SparkeyWriter(uri)
+          val it = iter.iterator
+          while (it.hasNext) {
+            val kv = it.next()
+            writer.put(kv._1.toString, kv._2.toString)
+          }
+          writer.close()
+          uri
+        }
+    }
+
+    /** Write the contents of this SCollection as a Sparkey file using the default uri. */
+    def asSparkey: SCollection[SparkeyUri] = {
+      val uri = ScioUtil.tempLocation(self.context.options) + "/sparkey"
+      this.asSparkey(SparkeyUri(uri))
+    }
+
+    def asSparkeySideInput: SideInput[SparkeyReader] = self.asSparkey.asSparkeySideInput()
+  }
+
+  implicit class SparkeySCollection(val self: SCollection[SparkeyUri]) {
+    def asSparkeySideInput(): SideInput[SparkeyReader] = {
+      val view = self.applyInternal(View.asSingleton())
+      new SparkeySideInput(view)
+    }
+  }
+
+  implicit class SparkeyPairSCollectionFunctions(val self: SCollection[(String, String)]) {
+    def asSparkeySideInput(): SideInput[SparkeyReader] = self.asSparkey.asSparkeySideInput()
+  }
+
+  implicit class RichSparkeyReader(val self: SparkeyReader) extends Map[String, String] {
+    override def get(key: String): Option[String] = Option(self.getAsString(key))
+
+    override def iterator: Iterator[(String, String)] = new Iterator[(String, String)] {
+      private val delegate = self.iterator()
+
+      override def hasNext: Boolean = delegate.hasNext
+
+      override def next(): (String, String) = {
+        val entry = delegate.next()
+        (entry.getKeyAsString, entry.getValueAsString)
+      }
+    }
+    //scalastyle:off method.name
+    override def +[B1 >: String](kv: (String, B1)): Map[String, B1] = ???
```

Comment:
`???` defaults to `new NotImplementedError()`
Maybe add an message "Not implemented. Sparkey-backed map."